File: CRM-Server/app/schemas/approval_ai.py
"""
AI 解析审批流程配置 Schema

用于 AI 辅助创建审批流程功能的请求和响应结构
"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from decimal import Decimal

from app.constants.business_types import BusinessType, is_valid_business_type
from app.schemas.approval import ApprovalNodeCreate


class ApprovalAIParsedNode(BaseModel):
    """AI 解析的审批节点"""
    node_name: str = Field(..., description="节点名称，如：部门经理审批")
    node_code: str = Field(..., description="节点编码，如：DEPT_MANAGER")
    node_order: int = Field(..., gt=0, description="审批顺序，从1开始")
    approve_role: str = Field(..., description="审批角色编码，如：SALES_DIRECTOR")
    description: Optional[str] = Field(None, description="节点描述")
    is_required: int = Field(1, ge=0, le=1, description="是否必须审批，默认1")


class ApprovalAIParsedFlow(BaseModel):
    """AI 解析的审批流程

    注意：min_amount/max_amount 使用 float 类型，
    确保 JSON 序列化兼容性（不需要 SSEJsonEncoder）
    """
    flow_name: str = Field(..., description="流程名称")
    flow_code: str = Field(..., description="流程编码")
    description: Optional[str] = Field(None, description="流程描述")
    # 改为 float 类型（不再使用 Decimal），确保 SSE JSON 序列化兼容
    min_amount: Optional[float] = Field(None, ge=0, description="最小金额（元）")
    max_amount: Optional[float] = Field(None, ge=0, description="最大金额（元）")
    license_type: Optional[str] = Field(None, description="授权类型，如：STANDARD")
    # B3：单据类型 CONTRACT/PAYMENT/INVOICE/LICENSE，默认 CONTRACT（兼容既有合同解析行为）
    business_type: str = Field(
        BusinessType.CONTRACT,
        description="流程适用单据类型：CONTRACT（合同，默认）/PAYMENT（回款）/INVOICE（发票）/LICENSE（License申请）",
    )
    nodes: List[ApprovalAIParsedNode] = Field(..., min_length=1, description="审批节点列表")

    @field_validator("business_type")
    @classmethod
    def validate_business_type(cls, v):
        """非法或缺失 business_type 回退 CONTRACT，保证模型始终有效。"""
        if not v or not is_valid_business_type(v):
            return BusinessType.CONTRACT
        return v

    def to_sse_dict(self) -> dict:
        """转换为 SSE 可序列化的字典

        使用 model_dump(mode='json') 确保所有类型兼容 JSON
        """
        return self.model_dump(mode='json')


class ApprovalAIParseRequest(BaseModel):
    """AI 解析请求"""
    content: str = Field(..., min_length=1, description="用户自然语言描述")


class ApprovalAIParseResponse(BaseModel):
    """AI 解析响应"""
    flow: ApprovalAIParsedFlow
    thinking_process: str = Field(..., description="AI 思考过程")


class ApprovalAICreateRequest(BaseModel):
    """创建审批流程请求（用户确认后）

    复用现有 ApprovalNodeCreate Schema，确保与 CRUD 接口兼容
    """
    flow_name: str = Field(..., description="流程名称")
    flow_code: str = Field(..., description="流程编码")
    description: Optional[str] = Field(None, description="流程描述")
    min_amount: Optional[Decimal] = Field(None, ge=0, description="最小金额（元）")
    max_amount: Optional[Decimal] = Field(None, ge=0, description="最大金额（元）")
    license_type: Optional[str] = Field(None, description="授权类型")
    # B3：落库 ApprovalFlow.business_type（A5 已加该列），默认 CONTRACT
    business_type: str = Field(
        BusinessType.CONTRACT,
        description="流程适用单据类型：CONTRACT/PAYMENT/INVOICE/LICENSE，默认 CONTRACT",
    )
    nodes: List[ApprovalNodeCreate] = Field(..., min_length=1, description="审批节点列表")

    @field_validator("business_type")
    @classmethod
    def validate_business_type(cls, v):
        if not v or not is_valid_business_type(v):
            return BusinessType.CONTRACT
        return v

    @field_validator('max_amount')
    @classmethod
    def validate_amount_range(cls, v, info):
        """可选校验：金额范围逻辑"""
        min_amount = info.data.get('min_amount')
        if v is not None and min_amount is not None and v <= min_amount:
            # 提示警告但不阻止（用户可能故意只设下限）
            pass
        return v

File: CRM-Client/src/api/approvalAI.ts
// CRM-Client/src/api/approvalAI.ts

/**
 * AI 审批流程解析 API
 */
import request from '@/utils/request'

/**
 * 预定义审批角色（匹配后端 constants/approval_roles.py）
 */
export const ALLOWED_APPROVAL_ROLES = [
  'TEAM_ADMIN',       // 团队所有者
  'SALES_DIRECTOR',   // 销售总监
  'FINANCE',          // 财务人员
  'SALES_MEMBER'      // 销售成员
] as const

export const ROLE_DISPLAY_NAMES: Record<string, string> = {
  TEAM_ADMIN: '团队所有者',
  SALES_DIRECTOR: '销售总监',
  FINANCE: '财务人员',
  SALES_MEMBER: '销售成员'
}

/**
 * AI 解析的审批节点（匹配 ApprovalAIParsedNode）
 */
export interface ApprovalAIParsedNode {
  node_name: string
  node_code: string
  node_order: number  // >= 1
  approve_role: string  // 预定义角色编码
  description?: string
  is_required: number  // 0 或 1，默认 1
}

/**
 * AI 解析的审批流程（匹配 ApprovalAIParsedFlow）
 */
export interface ApprovalAIParsedFlow {
  flow_name: string
  flow_code: string
  description?: string
  min_amount?: number  // Decimal → number
  max_amount?: number
  license_type?: string
  nodes: ApprovalAIParsedNode[]  // min_length=1
}

/**
 * 错误恢复动作（error 事件可附带的可操作按钮）
 */
export interface ErrorAction {
  type: 'retry' | 'simplify' | 'help'
  label: string
  suggestion?: string  // retry/simplify 时提供示例输入
  url?: string         // help 时提供跳转链接
}

/**
 * SSE 事件类型
 */
export interface ApprovalAIParseSSEEvent {
  event: 'status' | 'content' | 'parsed' | 'error' | 'done' | 'progress'
  message?: string
  content?: string
  flow?: ApprovalAIParsedFlow
  thinking_process?: string
  success?: boolean          // done 事件使用
  percentage?: number        // progress 事件使用
  stage?: 'analyzing' | 'generating' | 'validating'  // progress 阶段
  recovery?: string          // error 事件恢复提示
  actions?: ErrorAction[]    // error 事件可操作按钮
  estimated_seconds?: number // status 事件预估时间
  can_cancel?: boolean       // status 事件是否可取消
}

/**
 * 创建请求（使用 ApprovalNodeCreate 结构）
 */
export interface ApprovalAICreateRequest {
  flow_name: string
  flow_code: string
  description?: string
  min_amount?: number
  max_amount?: number
  license_type?: string
  nodes: ApprovalAIParsedNode[]  // 与 ApprovalNodeCreate 字段一致
}

/**
 * 创建响应
 */
export interface ApprovalAICreateResult {
  success: boolean
  message: string
  data: {
    flow: {
      id: number
      flow_name: string
      flow_code: string
      description?: string
      min_amount?: number
      max_amount?: number
      license_type?: string
    }
    nodes: {
      id: number
      node_name: string
      node_code: string
      node_order: number
      approve_role: string
    }[]
  }
}

export const approvalAiApi = {
  /**
   * AI 解析审批流程配置（SSE 流式响应）
   * @param data 解析请求数据
   * @param onEvent SSE 事件回调
   * @param token JWT token
   */
  parseSSE: async (
    data: { content: string },
    onEvent: (event: ApprovalAIParseSSEEvent) => void,
    token: string
  ): Promise<void> => {
    const url = '/api/v1/approval-ai/parse'

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify(data)
    })

    if (!response.ok) {
      throw new Error(`HTTP error: ${response.status}`)
    }

    const reader = response.body?.getReader()
    if (!reader) {
      throw new Error('No response body')
    }

    const decoder = new TextDecoder()
    let buffer = ''

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })

      // 解析 SSE 数据
      const lines = buffer.split('\n\n')
      const lastLine = lines.pop()
      buffer = lastLine !== undefined ? lastLine : ''

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const eventData = JSON.parse(line.slice(6)) as ApprovalAIParseSSEEvent
            onEvent(eventData)

            // 收到 parsed / error / done 事件后结束（done 为终端事件，含 success）
            if (eventData.event === 'parsed' || eventData.event === 'error' || eventData.event === 'done') {
              return
            }
          } catch {
            // 忽略解析错误
          }
        }
      }
    }
  },

  /**
   * 从 AI 解析结果创建审批流程
   * @param data 创建请求数据
   * @returns 创建结果
   */
  createFromAI: (data: ApprovalAICreateRequest): Promise<ApprovalAICreateResult> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    return request.post<ApprovalAICreateResult>(
      '/v1/approval-ai/create',
      data
    )
  }
}